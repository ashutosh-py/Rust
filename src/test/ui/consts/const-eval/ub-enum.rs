#[repr(usize)]
#[derive(Copy, Clone)]
enum Enum {
    A = 0,
}
union TransmuteEnum {
    a: &'static u8,
    b: Enum,
}

// A pointer is guaranteed non-null
const BAD_ENUM: Enum = unsafe { TransmuteEnum { a: &1 }.b };
//~^ ERROR this constant likely exhibits undefined behavior

// Invalid enum discriminant
#[repr(usize)]
#[derive(Copy, Clone)]
enum Enum2 {
    A = 2,
}
union TransmuteEnum2 {
    a: usize,
    b: Enum2,
}
const BAD_ENUM2 : Enum2 = unsafe { TransmuteEnum2 { a: 0 }.b };
//~^ ERROR this constant likely exhibits undefined behavior

// Invalid enum field content (mostly to test printing of apths for enum tuple
// variants and tuples).
union TransmuteChar {
    a: u32,
    b: char,
}
// Need to create something which does not clash with enum layout optimizations.
const BAD_ENUM_CHAR : Option<(char, char)> = Some(('x', unsafe { TransmuteChar { a: !0 }.b }));
//~^ ERROR this constant likely exhibits undefined behavior

fn main() {
}
