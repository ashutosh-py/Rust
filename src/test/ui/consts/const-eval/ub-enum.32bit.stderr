error[E0080]: evaluation of constant value failed
  --> $DIR/ub-enum.rs:24:33
   |
LL | const BAD_ENUM: Enum = unsafe { mem::transmute(1usize) };
   |                                 ^^^^^^^^^^^^^^^^^^^^^^ type validation failed at .<enum-tag>: encountered 0x00000001, but expected a valid enum tag

error[E0080]: evaluation of constant value failed
  --> $DIR/ub-enum.rs:27:37
   |
LL | const BAD_ENUM_PTR: Enum = unsafe { mem::transmute(&1) };
   |                                     ^^^^^^^^^^^^^^^^^^ type validation failed at .<enum-tag>: encountered pointer to alloc9, but expected plain (non-pointer) bytes

error[E0080]: evaluation of constant value failed
  --> $DIR/ub-enum.rs:30:47
   |
LL | const BAD_ENUM_WRAPPED: Wrap<Enum> = unsafe { mem::transmute(&1) };
   |                                               ^^^^^^^^^^^^^^^^^^ type validation failed at .0.<enum-tag>: encountered pointer to alloc13, but expected plain (non-pointer) bytes

error[E0080]: evaluation of constant value failed
  --> $DIR/ub-enum.rs:42:35
   |
LL | const BAD_ENUM2: Enum2 = unsafe { mem::transmute(0usize) };
   |                                   ^^^^^^^^^^^^^^^^^^^^^^ type validation failed at .<enum-tag>: encountered 0x00000000, but expected a valid enum tag

error[E0080]: evaluation of constant value failed
  --> $DIR/ub-enum.rs:44:39
   |
LL | const BAD_ENUM2_PTR: Enum2 = unsafe { mem::transmute(&0) };
   |                                       ^^^^^^^^^^^^^^^^^^ type validation failed at .<enum-tag>: encountered pointer to alloc19, but expected plain (non-pointer) bytes

error[E0080]: evaluation of constant value failed
  --> $DIR/ub-enum.rs:47:49
   |
LL | const BAD_ENUM2_WRAPPED: Wrap<Enum2> = unsafe { mem::transmute(&0) };
   |                                                 ^^^^^^^^^^^^^^^^^^ type validation failed at .0.<enum-tag>: encountered pointer to alloc23, but expected plain (non-pointer) bytes

error[E0080]: evaluation of constant value failed
  --> $DIR/ub-enum.rs:56:42
   |
LL | const BAD_ENUM2_UNDEF : Enum2 = unsafe { MaybeUninit { uninit: () }.init };
   |                                          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ type validation failed at .<enum-tag>: encountered uninitialized bytes, but expected initialized bytes

error[E0080]: evaluation of constant value failed
  --> $DIR/ub-enum.rs:60:54
   |
LL | const BAD_ENUM2_OPTION_PTR: Option<Enum2> = unsafe { mem::transmute(&0) };
   |                                                      ^^^^^^^^^^^^^^^^^^ type validation failed at .<enum-tag>: encountered pointer to alloc30, but expected plain (non-pointer) bytes

error[E0080]: evaluation of constant value failed
  --> $DIR/ub-enum.rs:77:62
   |
LL | const BAD_UNINHABITED_VARIANT1: UninhDiscriminant = unsafe { mem::transmute(1u8) };
   |                                                              ^^^^^^^^^^^^^^^^^^^ type validation failed at .<enum-variant(B)>.0: encountered a value of the never type `!`

error[E0080]: evaluation of constant value failed
  --> $DIR/ub-enum.rs:79:62
   |
LL | const BAD_UNINHABITED_VARIANT2: UninhDiscriminant = unsafe { mem::transmute(3u8) };
   |                                                              ^^^^^^^^^^^^^^^^^^^ type validation failed at .<enum-variant(D)>.0: encountered a value of uninhabited type Never

error[E0080]: evaluation of constant value failed
  --> $DIR/ub-enum.rs:87:67
   |
LL | const BAD_OPTION_CHAR: Option<(char, char)> = Some(('x', unsafe { mem::transmute(!0u32) }));
   |                                                                   ^^^^^^^^^^^^^^^^^^^^^ type validation failed: encountered 0xffffffff, but expected a valid unicode scalar value (in `0..=0x10FFFF` but not in `0xD800..=0xDFFF`)

error[E0080]: evaluation of constant value failed
  --> $DIR/ub-enum.rs:92:77
   |
LL | const BAD_UNINHABITED_WITH_DATA1: Result<(i32, Never), (i32, !)> = unsafe { mem::transmute(0u64) };
   |                                                                             ^^^^^^^^^^^^^^^^^^^^ transmuting to uninhabited type

error[E0080]: evaluation of constant value failed
  --> $DIR/ub-enum.rs:94:77
   |
LL | const BAD_UNINHABITED_WITH_DATA2: Result<(i32, !), (i32, Never)> = unsafe { mem::transmute(0u64) };
   |                                                                             ^^^^^^^^^^^^^^^^^^^^ transmuting to uninhabited type

error: aborting due to 13 previous errors

For more information about this error, try `rustc --explain E0080`.
